course5/homework04/homework04.cc

```diff
@@ -0,0 +1,47 @@
+#include "Stream.h"
+#include <cstdio>
+#include <iostream>
+#include <string>
+
+//Stream Cout = Stream(std::cout);
+Stream Cout = Stream(stdout);
+
+int main()
+{
+	
+	// Test catching
+	try
+	{
+		Stream s("./non-existent/output.txt");
+	}
+	catch (const std::exception& e)
+	{
+		std::cout << "Caught internal program error: " << e.what();
+		std::cout << "continuing for now... " << std::endl;
+		//std::exit(0);
+	}
+
+	try
+	{
+		std::cout << "Testing Stream object to output file output.txt " << std::endl; 
+		Stream s("output.txt");
+		std::string w = "World!"; 
+		s << "Hello, " << w; 
+		std::cout << "done" << std::endl; 
+
+		std::cout << "Testing Cout" << std::endl; 
+		Cout << "2 + 2 = "<< 2 + 2 << "\n"; 
+		std::cout << "done" << std::endl; 
+	
+		std::cout << "Testing Stream object with flush() to output file file.txt " << std::endl; 
+		Stream f("file.txt");
+		f << "hey" << Endl;
+		std::quick_exit(0);
+		
+	}
+	catch (const std::exception& e)
```

Comment:
It is a good practice not to catch exception on the top level, but rather to let if flow.
This way you'll not only get the error message, but more meta-data and a core dump, with a stack trace to the point when the exception was thrown.

We'll discuss it in more details.